File: solutions/javascript/allergies/1/allergies.js


export class Allergies {

  static PRODUCTS = ['eggs', 'peanuts', 'shellfish', 'strawberries',
    'tomatoes', 'chocolate', 'pollen', 'cats'];
  
  constructor(score) {
    // each specific allergy is a multiple of 2, converting the score to a binary number
    // and reversing so that the lowest digit is first in the list (eggs).
    this._allergyTests = score.toString(2).split('').reverse();
  }

  list() {
    const allergies = [];
    for (let i = 0; i < this._allergyTests.length && i < Allergies.PRODUCTS.length; i++) {
      if (this._allergyTests[i] == '1') allergies.push(Allergies.PRODUCTS[i]);
    }
    return allergies;
  }

  allergicTo(product) {
    const idx = Allergies.PRODUCTS.indexOf(product);
    if (idx == -1) return false;
    return this._allergyTests[idx] == '1';
  }
}

File: solutions/elixir/allergies/1/lib/allergies.ex
defmodule Allergies do

  import Bitwise

  @codes [eggs: 1, peanuts: 2, shellfish: 4, strawberries: 8, tomatoes: 16,
      chocolate: 32, pollen: 64, cats: 128]

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    Enum.reduce(@codes, [], fn ({allergy, code}, acc) -> 
        if((flags &&& code) != 0, do: [Atom.to_string(allergy) | acc], else: acc)
      end)
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    item in list(flags)
  end
end
